Annotations

EDsCODE opened the pull request:
## Changes

- implements #708 
- refactor linegraph.js to be a hook
- fix dashboard tests

## Checklist
- [ ] All querysets/queries filter by Team (if applicable)
- [ ] Backend tests (if applicable)
- [ ] Cypress E2E tests (if applicable)

EDsCODE committed:
draw a vertical line

EDsCODE committed:
some annotation

EDsCODE committed:
Merge branch 'master' into annotations

EDsCODE committed:
progress

EDsCODE committed:
progress on logic and mouse events

EDsCODE committed:
popover aligned

EDsCODE committed:
comment popup working

EDsCODE committed:
fix spacing

EDsCODE committed:
Merge branch 'master' into annotations

EDsCODE committed:
add model

EDsCODE committed:
annotation endpoint

EDsCODE committed:
initial working

EDsCODE committed:
frontend create dashbaord item with annotations working/ modified model

EDsCODE committed:
remove plugin

EDsCODE committed:
add sizing hook

EDsCODE committed:
fully working basic implementation

EDsCODE committed:
style the notes

EDsCODE committed:
annotations delete

EDsCODE committed:
annotations base complete

EDsCODE committed:
tweaking spacing and styling

EDsCODE committed:
enforce input limit

EDsCODE committed:
merge master

EDsCODE committed:
format

EDsCODE committed:
add margin

EDsCODE committed:
restrict deletes to proper users

EDsCODE committed:
modify styling

EDsCODE committed:
add api tests

EDsCODE committed:
remove console log from merge

EDsCODE commented:
@mariusandra One thing that's missing that just came to mind is the ability to add an annotation globally. Right now you add annotations specifically to a graph. I can add the functionality in the followup PR

mariusandra committed:
add error boundary to graph

mariusandra commented:
Hi! Overall this looks great! 

First some UX feedback and I'll check the code right after:

1. I think it would be great to show the year here and in the popup when adding an annotation. Right now it's obvious since all events ever recorded by PostHog are in 2020, but soon this might become problematic:

![image](https://user-images.githubusercontent.com/53387/86759881-1b35ab80-c045-11ea-95a9-8ecc0f0f0b3b.png)

2. I don't know how that happened, but the labels don't disappear anymore when you move out of the box:

![image](https://user-images.githubusercontent.com/53387/86760111-45876900-c045-11ea-8650-e051552e418b.png)

3. I think that if there are no annotations present, on smaller dashboards items the white space takes too much space:

![image](https://user-images.githubusercontent.com/53387/86760804-ca728280-c045-11ea-98e0-8e1732a8d4fb.png)

I don't know what's the best way to place them though. Would it look too bad if the "+" button would come over the dates?

4. I'm not sure why, but a lot of my dashboard items started to error when I switched to the `annotations` branch:

![image](https://user-images.githubusercontent.com/53387/86761358-32c16400-c046-11ea-813e-fddeb644686c.png)

They would properly crash the entire page, so I added an (antd) error boundary to at least limit the crashes to one item (a smart thing to do in any case). 

The error itself is:

```
Uncaught TypeError: Cannot read property '0' of undefined
    at Annotations (Annotations.js:17)
    at renderWithHooks (react-dom.development.js:14826)
    at updateFunctionComponent (react-dom.development.js:17059)
    at updateSimpleMemoComponent (react-dom.development.js:16997)
    at updateMemoComponent (react-dom.development.js:16886)
    at beginWork (react-dom.development.js:18716)
    at HTMLUnknownElement.callCallback (react-dom.development.js:189)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:238)
    at invokeGuardedCallback (react-dom.development.js:293)
    at beginWork$1 (react-dom.development.js:23235)
```

on this line: 

```js
const [diffType, setDiffType] = useState(determineDifferenceType(dates[0], dates[1]))
```

When I fixed it (or tried to) I got another error with `dates.forEach` below. Once I fixed that I got another one, this time inside LineChart... so I stopped whacking moles.

Interestingly, the charts that throw errors are all pie charts.

5. If I add an annotation on a dashboard, the message still shows this:

![image](https://user-images.githubusercontent.com/53387/86764508-60a7a800-c048-11ea-8bd7-9a21d6f5626c.png)

6. If I move over the "+" button, it moves chunkily (updates only onmouseout)... if I move under it, it catches every point... but then I can't just click to add an item:

![screencast 2020-07-07 11-53-49](https://user-images.githubusercontent.com/53387/86764813-a2d0e980-c048-11ea-9a15-6e65fb0058ce.gif)

7. Would be great to autofocus this textarea:

![image](https://user-images.githubusercontent.com/53387/86764973-d6137880-c048-11ea-80a4-27d3b7e8e4e8.png)

8. It might make sense to highlight the current day when hovering over the annotation for that day:

![screencast 2020-07-07 11-56-41](https://user-images.githubusercontent.com/53387/86765173-1b37aa80-c049-11ea-8af1-3b9eab2c3690.gif)
 
9. There's also z-index issue when I click on an existing annotation when I have the new one open:

![image](https://user-images.githubusercontent.com/53387/86765323-4de1a300-c049-11ea-9c2c-2a397e57b71e.png)

I think the popup for the new textarea should close then (but retain the text for if you click "+" again), like it does when you go the other way (click + after having some other list of annotation sopen)

10. The "blue" color makes sense on the white dashboard items, but I think the other ones (blue, green, etc) should get a different color for the annotation buttons as well. 

![image](https://user-images.githubusercontent.com/53387/86765625-add84980-c049-11ea-98f1-403b16f1db3f.png)

:)

mariusandra commented on frontend/src/scenes/trends/LineGraph.js:
```diff
@@ -209,14 +264,75 @@ export class LineGraph extends Component {
+    )
 }
+
+const map = (value, x1, y1, x2, y2) => Math.floor(((value - x1) * (y2 - x2)) / (y1 - x1) + x2)
```
This `map` could be renamed to something less commonly used.

mariusandra reviewed (commented):
Some comments below :)

EDsCODE committed:
fix code comments

EDsCODE committed:
fix lint

EDsCODE commented on frontend/src/lib/components/Annotations/AnnotationMarker.js:
```diff
@@ -0,0 +1,133 @@
+const { TextArea } = Input
+import { humanFriendlyDetailedTime } from '~/lib/utils'
+import { DeleteOutlined } from '@ant-design/icons'
+import _ from 'lodash'
```
#1159 added a ticket as the dependency appears in various places that we can just remove altogether in a separate PR

EDsCODE committed:
fix how name is determined

EDsCODE commented on frontend/src/lib/components/Annotations/AnnotationMarker.js:
```diff
@@ -0,0 +1,133 @@
+                    {label}
+                </Row>
+            }
+            {...{ ...(visible !== null && visible !== undefined && { visible }) }}
```
If the visible prop is set at all the component becomes a controlled state where it won't automatically handle visibility state. I only want this in the case where I pass `visible` from above and in some cases I don't have to. That horrid logic is so that the visible prop is only _touched_ in specific cases

mariusandra replied:
Might be wise to add a `/* comment */` explaining this to future generations

EDsCODE committed:
nit

EDsCODE committed:
refactor into logic

EDsCODE committed:
ux fixes

EDsCODE committed:
colors

EDsCODE commented:
@mariusandra I think I hit/responded to every point! Specifically, the UX changes that I couldn't fix:

3. it's really buggy to change the chart spacing dynamically at least in the attempts I've had. Placing the button above the axis or labels looks really wrong so this issue is still pending

6 and 8 are tricky fixes because the way I've created annotations is by adding a layer on top of the line graph and it's difficult to update styles related to the chart and even if we did it would rebuild the chart every time

Will continue tweaking to see if I can resolve any of the above issues especially 3 but otherwise good for another look

EDsCODE committed:
inting errors

EDsCODE committed:
Merge branch 'master' into annotations

EDsCODE committed:
fix dashbaord test

EDsCODE committed:
remove hardcoded test

EDsCODE committed:
fix linting again

EDsCODE committed:
remove autofocus

EDsCODE committed:
add dynamic spacing for annotations and fix margins

EDsCODE committed:
more patches

EDsCODE committed:
disable annotations on compared graph

mariusandra committed:
Merge branch 'annotations' of github.com:PostHog/posthog into annotations

mariusandra reviewed (commented):
Hey! I found some things to comment on regarding the code... and there are some UX points below.

Also, I had already added an `<ErrorBoundary />` to the dashboard items yesterday, but forgot to push it, so it's here now...

11. I think it should be without the gap on load... and only show the gap if there are annotations (slow gif, give it time):

![screencast 2020-07-08 09-47-43](https://user-images.githubusercontent.com/53387/86892254-4b8e5000-c100-11ea-81d3-7e8d6650b51f.gif)

12. This is not the friendliest way to let users interact with graphs. I'd assume 98% of the time when you hover over a graph you don't want to add an annotation, but just check the value at a point:

![screencast 2020-07-08 09-51-01](https://user-images.githubusercontent.com/53387/86892497-9e680780-c100-11ea-9f0a-8cc42daf29e6.gif)

Two ideas to solving it: 1) put the "+" over the dates, 2) add "add annotation" as an option in the menu, which then adds a gap and makes it possible to add something 3) like 2, but add it as a button next to the "reload" arrow and it could even work as a "toggle annotations" button.

13. I didn't test myself... but does this work on a phone and/or tablet? I still believe every feature should come out of the box with mobile support.

14. I still need to do some mouse gymnastics to add an annotation next to a day if the date are densely packed together. It's even hard to see that there are two buttons here. 

![screencast 2020-07-08 10-02-09](https://user-images.githubusercontent.com/53387/86893692-53e78a80-c102-11ea-83ee-b222e4cf76c1.gif)

Some thoughts for this: 1) Perhaps a `:hover { box-shadow: something }` style could help a bit to at least know what you're clicking? 2) The "+" and "1" buttons could have a slightly different style? 3) We could add "prev" and "next" buttons to the date (although this doesn't sound fun or easy, considering it should then move the tooltip as well left/right to coincide with the date)?

15. I'd also add the year to the "Add annotation" box

![image](https://user-images.githubusercontent.com/53387/86893826-81cccf00-c102-11ea-8f7e-b94d5b467e61.png)

16. I'm doubting if it's better to pause the "hover" effect while an existing annotation is open or not. Reason for pausing: consistency with the different tooltips (only one thing is open at a time)

![screencast 2020-07-08 10-13-44](https://user-images.githubusercontent.com/53387/86894883-fa805b00-c103-11ea-9f4f-c4f93e084f56.gif)

17. I'd also add a "X" to the tooltips... or at least the one with the existing annotations (the "add" one has a "cancel" button)

![image](https://user-images.githubusercontent.com/53387/86894951-15eb6600-c104-11ea-84c7-3680abb5b07e.png)

mariusandra commented:
Also, this should be rebased :)

mariusandra commented:
One more though regarding global annotations. Perhaps it would make sense to also add them in? At least I think they should be done before the release. It shouldn't be too hard to add either - just a checkbox when adding them should be enough UX wise I think?

EDsCODE committed:
merged master

EDsCODE committed:
code fixes

EDsCODE committed:
fix mouseover detail

EDsCODE committed:
add patch

EDsCODE commented:
Code comments addressed!

Managed to fix 6 from before so now the hover is very precise and should always reflect the right label. 
Addressed 14,15, 16, 17 (I made it so that you can't click away from the popup and have to explicitly close. lmk what you think about this)

**Remaining**
11 I think i fixed but I wasn't sure how to reproduce
12 I'd push back and say it's ok leaving it. I think the toggle should be included but maybe as a followup because as an initial feature push it'd be nice to really persuade people in using the annotations
13 Haven't tested this yet

**Global annotations**: also interested in doing this but I want to wrap this as a followup. UI is easy addition but it will take a whole other round of QA and testing to make sure the data is synced and playing nice with everything currently. For sake of clarity I'd like to do this in a separate PR whether we merge this one now or not

EDsCODE committed:
add control for mousedown

EDsCODE committed:
add condition

mariusandra commented:
18. Before the annotations load, the "+" is overlaid on the dates:

![screencast 2020-07-09 11-15-06](https://user-images.githubusercontent.com/53387/87021498-98902600-c1d5-11ea-96e1-f824a6ed56d3.gif)

19. This chart moves a bit to the right after the annotations are added... and consequently the annotations are off-center:
![screencast 2020-07-09 11-22-07](https://user-images.githubusercontent.com/53387/87022295-94183d00-c1d6-11ea-9e39-f7bdb6187247.gif)

20. Dragging issue:
![screencast 2020-07-09 11-32-16](https://user-images.githubusercontent.com/53387/87023384-ed34a080-c1d7-11ea-80c6-9977c84ec2c2.gif)

21. Would be nice to close the tooltip when pressing ESC. There's an `useEscapeKey` hook that should make this trivial.

22. (duplicate of 12) I'm still not terribly convinced that this is a nice transition. It feels odd having to wait 2 seconds for the dots to show up after I hover the graph. We might have bump this issue up to the court of public opinion (slack?). 

![screencast 2020-07-09 11-56-00](https://user-images.githubusercontent.com/53387/87025939-510c9880-c1db-11ea-9e11-531c303feb99.gif)

23. (duplicate of 13) I tested on mobile and it works more or less fine:
![20-07-09-12-08-10](https://user-images.githubusercontent.com/53387/87027556-9e8a0500-c1dd-11ea-9ff1-d8de4dcc361f.gif)

EDsCODE committed:
add escape key functionality

EDsCODE committed:
refine boundary recalculation condition

EDsCODE committed:
missing call

EDsCODE committed:
fix dragging listener

EDsCODE committed:
some refactoring and add control for dragging/popup

EDsCODE commented:
All should be addressed! for 22 I sped up the animation so it's more responsive lmk if that feels better otherwise yep we can move to slack for more input

EDsCODE committed:
better control when clustered

EDsCODE committed:
one more tweak

EDsCODE commented:
Note: error with class/id

EDsCODE committed:
use ref again

EDsCODE committed:
add condition

EDsCODE committed:
another condition

EDsCODE committed:
remove animations

timgl commented on frontend/src/scenes/trends/LineGraph.js:
```diff
@@ -1,54 +1,123 @@
+        const ticks = myLineChart.current.scales['x-axis-0'].ticks.length
+        const delta = rightExtent - leftExtent
+        const interval = delta / (ticks - 1)
+        const topExtent = myLineChart.current.scales['x-axis-0'].top + 12
```
so nitpicky but maybe just a touch less here, like `+ 8`

EDsCODE committed:
limit area

EDsCODE committed:
adjust spacing

EDsCODE committed:
condition

EDsCODE committed:
minor fix

EDsCODE committed:
colors

EDsCODE committed:
more patches

mariusandra commented:
Hey! I think the annotations part looks good and very usable now! 

The only thing I could find that's still broken is support for selecting text:

![2020-07-10 14 20 19](https://user-images.githubusercontent.com/53387/87154093-dfa31780-c2b8-11ea-885a-b9b030750e78.gif)
![2020-07-10 14 21 20](https://user-images.githubusercontent.com/53387/87154104-e3cf3500-c2b8-11ea-8124-3cc0c34058ef.gif)

EDsCODE commented:
is there a way to block dragging event on the dashboarditem? I opted to prioritize the dragging but seeing this text selection issue I think it would make more sense to block events going through the popup so dragging over the popup doesn't work

mariusandra commented:
There's this: https://github.com/PostHog/posthog/blob/93aebb9e7bac9150bd91607c8fba4708342ea4f1/frontend/src/scenes/dashboard/DashboardItems.js#L72

Adding here `.ant-tooltip` or whatever is the real class might do the trick?

EDsCODE committed:
use draggablecancel

EDsCODE commented:
Yup I think that does it

mariusandra reviewed (approved):
Perfect! 😄 

Approved! 👍 

(I guess the cypress tests will come later though :D)
